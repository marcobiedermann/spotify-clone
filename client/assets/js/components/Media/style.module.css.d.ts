declare const styles: {
  readonly media: string;
  readonly media__body: string;
  readonly media__object: string;
  readonly 'media__object--left': string;
  readonly 'media__object--right': string;
};
export = styles;
