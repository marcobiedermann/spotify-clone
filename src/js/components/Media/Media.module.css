.media {
  display: flex;
}

.media__body {
  flex-grow: 1;
}

.media__object {
  flex-shrink: 0;
}

.media__object--left {
  margin-right: 1em;
  order: -1;
}

.media__object--right {
  margin-left: 1em;
  order: 1;
}
