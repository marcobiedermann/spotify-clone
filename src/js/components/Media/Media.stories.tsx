import type { Meta, StoryObj } from '@storybook/react';
import Media, { MediaBody, MediaObject } from '.';

const meta: Meta<typeof Media> = {
  component: Media,
  render: () => (
    <Media>
      <MediaObject>Media Object</MediaObject>
      <MediaBody>Media Body</MediaBody>
    </Media>
  ),
};

type Story = StoryObj<typeof Media>;

const Default: Story = {};

const Left: Story = {
  render: () => (
    <Media>
      <MediaObject direction="left">Media Object</MediaObject>
      <MediaBody>Media Body</MediaBody>
    </Media>
  ),
};

const Right: Story = {
  render: () => (
    <Media>
      <MediaObject direction="right">Media Object</MediaObject>
      <MediaBody>Media Body</MediaBody>
    </Media>
  ),
};

export { Default, Left, Right };
export default meta;
