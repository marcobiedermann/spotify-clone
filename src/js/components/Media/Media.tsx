import clsx from 'clsx';
import React, { ReactNode } from 'react';
import styles from './Media.module.css';

export interface MediaBodyProps {
  children: ReactNode;
  className?: string;
}

export function MediaBody(props: MediaBodyProps): JSX.Element {
  const { className, ...otherProps } = props;

  return <div className={clsx(className, styles.media__body)} {...otherProps} />;
}

export interface MediaObjectProps {
  children: ReactNode;
  className?: string;
  direction?: 'left' | 'right';
}

export function MediaObject(props: MediaObjectProps): JSX.Element {
  const { className, direction = 'left', ...otherProps } = props;

  return (
    <div
      className={clsx(className, styles.media__object, {
        [styles[`media__object--${direction}`]]: direction,
      })}
      {...otherProps}
    />
  );
}

export interface MediaProps {
  children: ReactNode;
  className?: string;
}

function Media(props: MediaProps): JSX.Element {
  const { className, ...otherProps } = props;

  return <div className={clsx(className, styles.media)} {...otherProps} />;
}

export default Media;
