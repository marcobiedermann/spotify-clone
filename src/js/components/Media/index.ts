import Media from './Media';

export default Media;
export * from './Media';
