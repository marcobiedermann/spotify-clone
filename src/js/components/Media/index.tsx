import classNames from 'classnames';
import React, { ReactNode } from 'react';
import styles from './style.module.css';

export interface MediaBodyProps {
  children: ReactNode;
  className?: string;
}

export function MediaBody(props: MediaBodyProps): JSX.Element {
  const { className, ...otherProps } = props;

  return <div className={classNames(className, styles.media__body)} {...otherProps} />;
}

export interface MediaObjectProps {
  children: ReactNode;
  className?: string;
  direction?: 'left' | 'right';
}

export function MediaObject(props: MediaObjectProps): JSX.Element {
  const { className, direction = 'left', ...otherProps } = props;

  return (
    <div
      className={classNames(className, styles.media__object, {
        [`media__object--${direction}`]: direction,
      })}
      {...otherProps}
    />
  );
}

export interface MediaProps {
  children: ReactNode;
  className?: string;
}

function Media(props: MediaProps): JSX.Element {
  const { className, ...otherProps } = props;

  return <div className={classNames(className, styles.media)} {...otherProps} />;
}

export default Media;
