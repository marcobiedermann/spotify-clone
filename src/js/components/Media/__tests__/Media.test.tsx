import { render } from '@testing-library/react';
import React from 'react';
import { describe, expect, it } from 'vitest';
import Media, { MediaBody, MediaObject } from '..';

describe('Media Component', () => {
  it('should render MediaObject with default direction correctly', () => {
    expect.assertions(1);

    const { container } = render(
      <Media>
        <MediaObject>Media Object</MediaObject>
        <MediaBody>Media Body</MediaBody>
      </Media>,
    );

    expect(container.firstChild).toMatchInlineSnapshot(`
      <div
        class="media"
      >
        <div
          class="media__object media__object--left"
        >
          Media Object
        </div>
        <div
          class="media__body"
        >
          Media Body
        </div>
      </div>
    `);
  });

  it('should render MediaObject with direction `left` correctly', () => {
    expect.assertions(1);

    const { container } = render(
      <Media>
        <MediaObject direction="left">Media Object</MediaObject>
        <MediaBody>Media Body</MediaBody>
      </Media>,
    );

    expect(container.firstChild).toMatchInlineSnapshot(`
      <div
        class="media"
      >
        <div
          class="media__object media__object--left"
        >
          Media Object
        </div>
        <div
          class="media__body"
        >
          Media Body
        </div>
      </div>
    `);
  });

  it('should render MediaObject with direction `right` correctly', () => {
    expect.assertions(1);

    const { container } = render(
      <Media>
        <MediaObject direction="right">Media Object</MediaObject>
        <MediaBody>Media Body</MediaBody>
      </Media>,
    );

    expect(container.firstChild).toMatchInlineSnapshot(`
      <div
        class="media"
      >
        <div
          class="media__object media__object--right"
        >
          Media Object
        </div>
        <div
          class="media__body"
        >
          Media Body
        </div>
      </div>
    `);
  });
});
