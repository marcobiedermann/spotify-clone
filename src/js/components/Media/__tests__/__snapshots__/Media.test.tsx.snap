// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Media Component > should render MediaObject with default direction correctly 1`] = `
<div
  class="media"
>
  <div
    class="media__object media__object--left"
  >
    Media Object
  </div>
  <div
    class="media__body"
  >
    Media Body
  </div>
</div>
`;

exports[`Media Component > should render MediaObject with direction \`left\` correctly 1`] = `
<div
  class="media"
>
  <div
    class="media__object media__object--left"
  >
    Media Object
  </div>
  <div
    class="media__body"
  >
    Media Body
  </div>
</div>
`;

exports[`Media Component > should render MediaObject with direction \`right\` correctly 1`] = `
<div
  class="media"
>
  <div
    class="media__object media__object--right"
  >
    Media Object
  </div>
  <div
    class="media__body"
  >
    Media Body
  </div>
</div>
`;
